lang/src/org/partiql/lang/eval/ExprValueExtensions.kt

```diff
@@ -329,43 +345,38 @@ fun ExprValue.cast(
                             val time = timeValue()
-                            val precision = when {
-                                // precisionProvided > MAX_PRECISION_FOR_TIME indicates that the precision was not specified.
-                                precisionProvided > MAX_PRECISION_FOR_TIME -> time.precision
-                                else -> precisionProvided
-                            }
-                            val tzMinutes = if (timezone == 1) {
-                                time.zoneOffset?.totalSeconds?.div(SECONDS_PER_MINUTE) ?: LOCAL_TIMEZONE_OFFSET.totalSeconds / SECONDS_PER_MINUTE
-                            } else {
-                                null
-                            }
-                            return valueFactory.newTime(Time.of(time.localTime.hour, time.localTime.minute, time.localTime.second, time.localTime.nano, precision, tzMinutes))
+                            return valueFactory.newTime(
+                                Time.of(
+                                    time.localTime.hour,
+                                    time.localTime.minute,
+                                    time.localTime.second,
+                                    time.localTime.nano,
+                                    precision?: time.precision,
+                                    when (targetSqlDataType) {
+                                        SqlDataType.TIME_WITH_TIME_ZONE -> time.zoneOffset?.totalSeconds?.div(SECONDS_PER_MINUTE)?: LOCAL_TIMEZONE_OFFSET.totalSeconds / SECONDS_PER_MINUTE
+                                        else -> null
+                                    }
+                                ))
                         }
                         type == TIMESTAMP -> {
                             val ts = timestampValue()
-                            val precision = when {
-                                precisionProvided > MAX_PRECISION_FOR_TIME -> ts.decimalSecond.scale()
-                                else -> precisionProvided
-                            }
                             return valueFactory.newTime(
                                 ts.hour,
                                 ts.minute,
                                 ts.second,
                                 (ts.decimalSecond.remainder(BigDecimal.ONE) * NANOS_PER_SECOND).toInt(),
-                                precision,
-                                if (timezone == 1) {
-                                    ts.localOffset?: castFailedErr(
+                                precision?: ts.decimalSecond.scale(),
+                                when (targetSqlDataType) {
+                                    SqlDataType.TIME_WITH_TIME_ZONE -> ts.localOffset?: castFailedErr(
                                         "Can't convert timestamp value with undefined time zone to TIME WITH TIME ZONE.",
                                         internal = false
                                     )
```

Comment:
Isn't this case `-00:00`, is that not supported?  If not, we should put a more meaningful error message.

Reply:
I see. I will change the error message. [This](https://www.javadoc.io/doc/com.amazon.ion/ion-java/latest/com/amazon/ion/Timestamp.html#getLocalOffset--) is the documentation for the `getLocalOffset` of `ion Timestamp`.
Would this error message work - 
```
"Can't convert timestamp value with unknown local offset to TIME WITH TIME ZONE.",
```